### PR TITLE
add method to retrieve instrumentation configuration by name

### PR DESCRIPTION
**Pull request by zeitlinger**
Follow up to https://github.com/open-telemetry/opentelemetry-java/pull/7923

See https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/15641#discussion_r2625857580

### CODE REVIEW

File: api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java [hunk 1]
```diff
@@ -30,6 +32,22 @@ public interface ConfigProvider {
...
 
+  /**
+   * Returns the {@link DeclarativeConfigProperties} for a specific instrumentation by name. If no
+   * configuration is available for the given name, an empty {@link DeclarativeConfigProperties} is
+   * returned.
+   *
+   * @param name the name of the instrumentation
+   * @return the {@link DeclarativeConfigProperties} for the given instrumentation name
+   */
+  default DeclarativeConfigProperties get(String name) {
```
Comment (jack-berg) [inline api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java]: I'm a bit worried that if `ConfigProvider` is expanded in the future to have other functions, it won't be clear that `get` is sugar for `getInstrumentationConfig`.

Can remove this risk with naming `getJavaInstrumentationConfig(String name)`, which I think is more clear that this method is about accessing a particular java instrumentation library's config.

Looking at this, I wonder if we should scrap #7920 in favor of an approach like this PR @trask.

This approach is a bit of a "have your cake and eat it to", since it still allows users to determine if `.instrumentation/development` was set, but provides syntactic sugar for the common case where you don't actually care.

File: api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java [hunk 2]
```diff
@@ -30,6 +32,22 @@ public interface ConfigProvider {
...
 
+  /**
+   * Returns the {@link DeclarativeConfigProperties} for a specific instrumentation by name. If no
+   * configuration is available for the given name, an empty {@link DeclarativeConfigProperties} is
+   * returned.
+   *
+   * @param name the name of the instrumentation
+   * @return the {@link DeclarativeConfigProperties} for the given instrumentation name
+   */
+  default DeclarativeConfigProperties getJavaInstrumentationConfig(String name) {
```
Comment (jack-berg) [inline api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java]: In the java SIG today we discussed adding an additional helper `DeclarativeConfigProperties getGeneralInstrumentationConfig()` which would operate the same but access the `.instrumentation/development.general` node.

Your call whether we do that in a separate PR or this one.

File: sdk-extensions/autoconfigure/src/testIncubating/java/io/opentelemetry/sdk/autoconfigure/DeclarativeConfigurationTest.java [hunk 3]
```diff
@@ -231,5 +231,9 @@ void configFile_ConfigProvider() {
     assertThat(InstrumentationConfigUtil.javaInstrumentationConfig(globalConfigProvider, "example"))
         .isNotNull()
         .satisfies(exampleConfig -> assertThat(exampleConfig.getString("key")).isEqualTo("value"));
+
+    // shortcut to get specific instrumentation config
+    assertThat(globalConfigProvider.getJavaInstrumentationConfig("example").getString("key"))
+        .isEqualTo("value");
```
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/testIncubating/java/io/opentelemetry/sdk/autoconfigure/DeclarativeConfigurationTest.java:230]: I think we need test coverage to verify that even for a noop config where `getInstrumentationConfig() == null`, `getJavaInstrumentationConfig(String)` still returns empty.

Maybe add in [ConfigProviderTest](https://github.com/open-telemetry/opentelemetry-java/blob/main/api/incubator/src/test/java/io/opentelemetry/api/incubator/ConfigProviderTest.java) since this test is concerned with higher level things.

File: api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java (hunk 2 above)
Comment (trask) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java]: Would be nice to also have a getDistributionConfig() to access the `.distribution` node (without null worries)

Comment (jack-berg) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java]: That's a different story. Right now the scope of what `ConfigProvider` knows about is limited to the `.instrumentation/development` node. Would need to have something like https://github.com/open-telemetry/opentelemetry-specification/pull/4800 to expand the scope. 

Note its still possible for distributions to access `.distribution` without PR#4800 with a pattern like:

```
OpenTelemetryConfigurationModel model = parse(new File(System.getEnv("OTEL_EXPERIMENTAL_CONFIG_FILE")));
DistributionModel distribution = model.getDistribution();
ExtendedOpenTelemetrySdk sdk = create(model);
```

Comment (zeitlinger) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java]: that pattern is brittle, e.g. doesn't work with the spring starter

**Comment by zeitlinger** [comment]
@jack-berg added the second method for "general" and improved test coverage

File: api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java [hunk 4]
```diff
@@ -30,6 +32,33 @@ public interface ConfigProvider {
...
+  /**
+   * Returns the {@link DeclarativeConfigProperties} for general instrumentation config by name. If
+   * no configuration is available for the given name, an empty {@link DeclarativeConfigProperties}
+   * is returned.
+   *
+   * @param name the name of the general instrumentation config
+   * @return the {@link DeclarativeConfigProperties} for the given general instrumentation config
+   *     name
+   */
+  default DeclarativeConfigProperties getGeneralInstrumentationConfig(String name) {
```
Comment (jack-berg) [inline api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java]: This works because for now there are no top level properties under `.instrumentation/development.general`. But if there were, callers would be unable to access them. I think we should omit the `String name` parameter for now and just return the `.instrumentation/development.general` node.

File: api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java (hunk 2 above)
Comment (trask) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java]: @zeitlinger in spring starter can't we get the distribution node from https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/dcee1b769f353bee881ab7be8c7caa7715ea5143/instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/OpenTelemetryAutoConfiguration.java#L158?

File: api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java (hunk 4 above)
Comment (trask) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java]: oh good point, can `.instrumentation/development.java.` have top-level properties?

I really don't mind `getJavaInstrumentationConfig().get(name)`

Comment (zeitlinger) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java]: This is only a shortcut method - so I think we're allowed to nudge towards what we think is best practice. 
I think that is the currently implemented solution, but I can live with either.

Comment (jack-berg) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java]: > oh good point, can .instrumentation/development.java. have top-level properties?

No - its properties are limited to objects: https://github.com/open-telemetry/opentelemetry-configuration/blob/main/schema-docs.md#experimentallanguagespecificinstrumentation-

In contrast, `.general` has explicit properties of `http` and `peer` which are objects, but nothing stating all properties are objects: https://github.com/open-telemetry/opentelemetry-configuration/blob/main/schema-docs.md#experimentalgeneralinstrumentation-

Comment (zeitlinger) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java]: ok, I'll adapt "general" then

Comment (zeitlinger) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java]: done

File: api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java (hunk 2 above)
Comment (zeitlinger) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java]: Good point, that would work for spring applications.

- `AgentListener` passes `AutoConfiguredOpenTelemetrySdk` - and that gives you access to `ConfigProvider` only - but Jacks suggestion would work
- if the user also checks for the corresponding system property

File: api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java [hunk 5]
```diff
@@ -30,6 +32,31 @@ public interface ConfigProvider {
...
+  }
+
+  /**
+   * Returns the {@link DeclarativeConfigProperties} for general instrumentation configuration. If
+   * the general configuration is not available, an empty {@link DeclarativeConfigProperties} is
+   * returned.
+   *
+   * @return the {@link DeclarativeConfigProperties} for the general instrumentation configuration
+   */
+  default DeclarativeConfigProperties getGeneralInstrumentationConfig() {
```
Comment (trask) [inline api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java:64]: would we be comfortable with something shorter, e.g.

- getInstrumentationGeneral()
- getInstrumentationJava(name)

Comment (zeitlinger) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java:64]: What about "getGeneralInstrumentation"?

Comment (jack-berg) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java:64]: > What about "getGeneralInstrumentation"?

I like this for general, but symmetric naming yields `getJavaInstrumentation(name)`, which makes it sound like its accessing the instrumentation rather than the config for the instrumentation. 

Not that `getInstrumentationJava(name)` is especially obvious either, but at least the method name reflects the path to the YAML node being accessed.

Comment (trask) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java:64]: maybe

- getInstrumentation(name)
- getGeneralInstrumentation()

with the idea that the "Java" is redundant for the Java SDK

Comment (jack-berg) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java:64]: Is getInstrumentation(name) too close to getInstrumentationConfig() given the difference in function?

Comment (trask) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java:64]: is there a way we could remove getInstrumentationConfig()?

what if we added something like `.exists()` to `DeclarativeConfigurationProperties` for the few people who may care about the difference between intermediate nodes being present or not

Comment (jack-berg) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java:64]: I'm not particularly attached to it, but there's the bit about it being part of the spec. Of course the spec can be changed. But the path of least resistance for the short term is to find method names that jive with the current spec and which can be characterized as syntactic sugar. The spec doesn't prohibit syntactic sugar so it's the sweet spot where we can improve usability quickly.

Comment (jack-berg) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java]: I'm not particularly attached to it, but there's the bit about it being part of the spec. Of course the spec can be changed. But the path of least resistance for the short term is to find method names that jive with the current spec and which can be characterized as syntactic sugar. The spec doesn't prohibit syntactic sugar so it's the sweet spot where we can improve usability quickly.

Comment (zeitlinger) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/config/ConfigProvider.java:64]: I went for 

- `getInstrumentationConfig`
- `getGeneralInstrumentationConfig`

now based on the feedback. Let me know what you think.

File: api/incubator/src/test/java/io/opentelemetry/api/incubator/ConfigProviderTest.java [hunk 6]
```diff
@@ -17,4 +17,12 @@ void noopEquality() {
     ConfigProvider noop = ConfigProvider.noop();
     assertThat(ConfigProvider.noop()).isSameAs(noop);
   }
+
+  @Test
+  void instrumentationConfigFallback() {
+    ConfigProvider configProvider = ConfigProvider.noop();
+    assertThat(configProvider.getInstrumentationConfig()).isNull();
+    assertThat(configProvider.getInstrumentationConfig("servlet")).isNotNull();
```
Comment (jkwatson) [inline api/incubator/src/test/java/io/opentelemetry/api/incubator/ConfigProviderTest.java]: this is surprising to me. I don't like that one of these is null and one of them is not. We should have a non-surprising API for this, IMO.

**Comment by zeitlinger** [comment]
@jack-berg I've moved the methods to ExtendedOpenTelemetry because 

- the comment about symetry: https://github.com/open-telemetry/opentelemetry-java/pull/7927#discussion_r2649309534
  - if both methods called `getInstrumentationConfig` are not in the same class, it will be less confusing that one of them is nullable and the other one is not
- considering that `getConfigProvider()` will become nullable when stabilized
  -  when stabilized, `getConfigProvider()` will move to `OpenTelemetry`
  - if declarative config is not used, `getConfigProvider()` has to return `null` or an empty object
  - an empty object doesn't can't be used to determine if declarative config is used or not, hence I think this is not viable ([getInstrumentationConfig()](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/api.md#get-instrumentation-config) only allows to determine if the `.instrumentation` node is set)

File: api/incubator/src/test/java/io/opentelemetry/api/incubator/ConfigProviderTest.java (hunk 6 above)
Comment (zeitlinger) [reply api/incubator/src/test/java/io/opentelemetry/api/incubator/ConfigProviderTest.java]: good point - addressed in https://github.com/open-telemetry/opentelemetry-java/pull/7927#issuecomment-3709499434

File: api/incubator/src/main/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetry.java [hunk 7]
```diff
@@ -5,13 +5,43 @@
...
+
+  /**
+   * Returns the {@link DeclarativeConfigProperties} for a specific instrumentation by name. If no
+   * configuration is available for the given name, an empty {@link DeclarativeConfigProperties} is
+   * returned.
+   *
+   * @param name the name of the instrumentation
+   * @return the {@link DeclarativeConfigProperties} for the given instrumentation name
+   */
+  default DeclarativeConfigProperties getInstrumentationConfig(String name) {
```
Comment (trask) [inline api/incubator/src/main/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetry.java:35]: I like these shortcuts on OpenTelemetry. Not sure about removing them from `ConfigProvider`. If there's agreement(?) on

> getConfigProvider() will be nullable once stability is reached

and `getInstrumentationConfig()` becoming non-null, why not do it now?

File: api/incubator/src/test/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetryTest.java [hunk 8]
```diff
@@ -70,4 +77,50 @@ private static void assertIsExtended(OpenTelemetry openTelemetry) {
...
+  void instrumentationConfigFallback() {
+    ConfigProvider configProvider = ConfigProvider.noop();
+    // cannot create ExtendedOpenTelemetry with noop ConfigProvider right now,
+    // but will be possible once stable API is available
+    ExtendedOpenTelemetry openTelemetry = spy(ExtendedOpenTelemetry.class);
+    when(openTelemetry.getConfigProvider()).thenReturn(configProvider);
+
+    assertThat(configProvider.getInstrumentationConfig()).isNull();
+    assertThat(openTelemetry.getInstrumentationConfig("servlet")).isNotNull();
+    assertThat(openTelemetry.getGeneralInstrumentationConfig()).isNotNull();
```
Comment (jkwatson) [inline api/incubator/src/test/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetryTest.java]: this is still a quite surprising set of responses. We should really make these consistent, one way or another.

**Review by jkwatson — CHANGES_REQUESTED** [review]

File: api/incubator/src/test/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetryTest.java (hunk 8 above)
Comment (trask) [reply api/incubator/src/test/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetryTest.java]: Usability is quite bad if they're nullable

@jack-berg can we make `getInstrumentationConfig()` non-null?

**Comment by zeitlinger** [comment]
Just discussed that in the config SIG:

* ConfigProvider is also meant to be used when declarative config is not used  
  * But maybe we don’t need to spec out how env variables are mapped \- unless other languages (than Java) would benefit from it  
* Could delete this sentence \- as it doesn’t seem to be needed based on Java experience  
  * *If the .instrumentation node is not set, get instrumentation config MUST return nil, null, undefined or another language-specific idiomatic pattern denoting empty.*  
* Plan for Java   
  * Move default methods back to ConfigProvider (or both)
  * SDK has to supply env var based implementation of ConfigProvider  
     * But that has java agent specific rules in there right now  
     * Could create a stripped down version for the SDK repo

File: api/incubator/src/test/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetryTest.java (hunk 8 above)
Comment (zeitlinger) [reply api/incubator/src/test/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetryTest.java]: based on the discussion in the config SIG, it sounds like making all non-null is the way to go - though we may need to update the spec first

File: api/incubator/src/main/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetry.java (hunk 7 above)
Comment (zeitlinger) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetry.java:35]: see https://github.com/open-telemetry/opentelemetry-java/pull/7927#issuecomment-3711456101

File: api/incubator/src/test/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetryTest.java (hunk 8 above)
Comment (trask) [reply api/incubator/src/test/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetryTest.java]: if there's agreement, I don't think we technically have to wait for spec change since this is in incubator module anyways. might be nice to get this into Friday's release.

Comment (zeitlinger) [reply api/incubator/src/test/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetryTest.java]: good idea: https://github.com/open-telemetry/opentelemetry-java/pull/7954

File: api/incubator/src/main/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetry.java (hunk 7 above)
Comment (zeitlinger) [reply api/incubator/src/main/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetry.java:35]: > I like these shortcuts on OpenTelemetry. Not sure about removing them from `ConfigProvider`

I have both now

**Comment by zeitlinger** [comment]
@trask @jkwatson @jack-berg https://github.com/open-telemetry/opentelemetry-java/pull/7954 is merged - this PR should be ready to be merged

File: api/incubator/src/main/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetry.java [hunk 9]
```diff
@@ -7,11 +7,35 @@
...
   /** Returns the {@link ConfigProvider} for this {@link OpenTelemetry}. */
   default ConfigProvider getConfigProvider() {
     return ConfigProvider.noop();
   }
+
+  /**
+   * Returns the {@link DeclarativeConfigProperties} for a specific instrumentation by name. If no
+   * configuration is available for the given name, an empty {@link DeclarativeConfigProperties} is
+   * returned.
+   *
```
Comment (trask) [inline api/incubator/src/main/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetry.java:23]:
```suggestion
   *
   * <p>For example, {@code getInstrumentationConfig("foo")} returns the node:
   *
   * <pre>{@code
   * instrumentation/development:
   *   java:
   *     foo:
   * }</pre>
   *
```

File: api/incubator/src/main/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetry.java [hunk 10]
```diff
@@ -7,11 +7,35 @@
...
+   * @param name the name of the instrumentation
+   * @return the {@link DeclarativeConfigProperties} for the given instrumentation name
+   */
+  default DeclarativeConfigProperties getInstrumentationConfig(String name) {
+    return getConfigProvider().getInstrumentationConfig(name);
+  }
+
+  /**
+   * Returns the {@link DeclarativeConfigProperties} for general instrumentation configuration. If
+   * the general configuration is not available, an empty {@link DeclarativeConfigProperties} is
```
Comment (trask) [inline api/incubator/src/main/java/io/opentelemetry/api/incubator/ExtendedOpenTelemetry.java]:
```suggestion
   * Returns the {@link DeclarativeConfigProperties} for general instrumentation configuration:
   *
   * <pre>{@code
   * instrumentation/development:
   *   general:
   * }</pre>
   *
   * If the general configuration is not available, an empty {@link DeclarativeConfigProperties} is
```

**Review by trask — APPROVED** [review]
Only have javadoc thoughts on how to make it clearer which node you're getting back.

**Comment by jack-berg** [comment]
> SDK has to supply env var based implementation of ConfigProvider 

I don't think this is true. The agent definitely needs an implementation of ConfigProvider based on env vars, but I think we could reasonably take the position that we don't aim to further improve the env var based config story, and thus won't provide a vanilla SDK (i.e. no agent) implementation of ConfigProvider based on env vars.

**Comment by zeitlinger** [comment]
> > SDK has to supply env var based implementation of ConfigProvider
> 
> I don't think this is true. The agent definitely needs an implementation of ConfigProvider based on env vars, but I think we could reasonably take the position that we don't aim to further improve the env var based config story, and thus won't provide a vanilla SDK (i.e. no agent) implementation of ConfigProvider based on env vars.

created https://github.com/open-telemetry/opentelemetry-java/issues/7961 to discuss this

**Comment by zeitlinger** [comment]
@jack-berg should be good to go now
